### ISSUE TITLE
Population of fields with the same "flat" name

### ISSUE THREAD

**Issue by dirchev** [Automattic/mongoose#6839]
**Do you want to request a *feature* or report a *bug*?**
BUG

**What is the current behavior?**

This was working on mongoose 4 and the bug is introduced in mongoose 5. 
When we have a schema with fields of type `ObjectId`, what share the same object key name - there is a population bug.

For example a schema looking like:

```javascript
{
  name: String,
  child: {type: mongoose.Schema.Types.ObjectId, ref: 'Child'},
  children: [
    {
      child: {type: mongoose.Schema.Types.ObjectId, ref: 'Child'},
      addedOn: { type: Date, default: Date.now }
    }
  ]
}
```

We have to populate `child` and `children.child`. Unfortunately, sometimes populate does not work as expected, populating the `children.child` to `null`.

**If the current behavior is a bug, please provide the steps to reproduce.**

```
var mongoose = require('mongoose')

describe('mongoose 5 double populate bug', function () {
  before(test.connectDB)
  after(test.disconnectDB)
  // MODELS
  var Child
  var Parent

  // DOCUMENTS
  var child
  var parent

  // POPULATE
  var parentFieldsToPopulate = [
    {path: 'children.child'},
    {path: 'child'}
  ]

  before(function () {
    var struct = {name: String}
    var childSchema = new mongoose.Schema(struct)
    Child = mongoose.model('Child', childSchema)
  })

  before(function () {
    var struct = {
      name: String,
      child: {type: mongoose.Schema.Types.ObjectId, ref: 'Child'},
      children: [
        {
          child: {type: mongoose.Schema.Types.ObjectId, ref: 'Child'},
          addedOn: { type: Date, default: Date.now }
        }
      ]
    }
    var parentSchema = new mongoose.Schema(struct)
    Parent = mongoose.model('Parent', parentSchema)
  })

  it('creates a child document', function (next) {
    child = new Child({
      name: 'Test Child'
    })
    child.save(next)
  })

  it('creates a parent document', function (next) {
    parent = new Parent({
      name: 'Test Parent',
      child: child.id
    })
    parent.save(next)
  })

  it('[does not work] adds child to parent', function (next) {
    Child.findById(child.id).exec(function (err, child) {
      if (err) return next(err)
      Parent.findById(parent.id).populate(parentFieldsToPopulate).exec(function (err, parent) {
        if (err) return next(err)
        parent.children.push({child: child})
        parent.save(function (err, parent) {
          if(err) return next(err)
          parent.populate(parentFieldsToPopulate, function (err, parent) {
            if (err) return next(err)
            expect(parent.children[0].child).to.exist
            next()
          })
        })
      })
    })
  })

  xit('[works] adds child to parent', function (next) {
    Child.findById(child.id).exec(function (err, child) {
      if (err) return next(err)
      Parent.findById(parent.id).populate(parentFieldsToPopulate).exec(function (err, parent) {
        if (err) return next(err)
        parent.children.push({child: child})
        parent.save(function (err, parent) {
          if(err) return next(err)
          Parent.findById(parent.id).populate(parentFieldsToPopulate).exec(function (err, parent) {
            if (err) return next(err)
            expect(parent.children[0].child).to.exist
            next()
          })
        })
      })
    })
  })
})
```

**What is the expected behaviour?**

The expected behaviour is to have `children.child` populated properly. The population works fine if I use `Model.populate` instead of `document.populate` (see the `[works]` test)

**Please mention your node.js, mongoose and MongoDB version.**

NodeJS: v8.11.3
mongoose: v5.2.7
MongoDB: v3.4.16

**Comment by dirchev** [comment]
Clarification:

The problem is when we `save` and `populate` a document, that has already been populated. So the test

```
it('adds child to parent', function (next) {
  Child.findById(child.id).exec(function (err, child) {
    if (err) return next(err)
    Parent.findById(parent.id)/* .populate(parentFieldsToPopulate) */.exec(function (err, parent) {
      if (err) return next(err)
      parent.children.push({child: child})
      parent.save(function (err, parent) {
        if(err) return next(err)
        parent.populate(parentFieldsToPopulate, function (err, parent) {
          if (err) return next(err)
          expect(parent.children[0].child).to.exist
          next()
        })
      })
    })
  })
})
```

works as expected

**Comment by vkarpov15** [comment]
Thanks for reporting, will investigate ASAP

**Comment by vkarpov15** [comment]
Looks like this was fixed in Mongoose 5.2.11, your test fails on 5.2.7 but succeeds on `>= 5.2.11`. It looks like the fix for #6908 also fixed this one.

**Comment by vkarpov15** [comment]
Looking a little closer, it turns out that 5.2.11 doesn't overwrite `children.child` with `null` but it still wouldn't populate `children.child`. I added a fix for this that will be in 5.2.14.
